### ISSUE TITLE
[Bug] Kompile pipeline duplicates modules

### ISSUE THREAD

**Issue by radumereuta** [runtimeverification/k#1838]
Git revision: 061c0ba (latest master)

Replicate with just command line:
```
radu@radu:~/work/k/k-distribution/tests/regression-new/kprove-haskell$ kompile verif.k --backend haskell
radu@radu:~/work/k/k-distribution/tests/regression-new/kprove-haskell$ kprove sum-spec.k --save-proof-definition-to save
#Top
radu@radu:~/work/k/k-distribution/tests/regression-new/kprove-haskell$ kprove test-spec.k -d save
[Error] Internal: Uncaught exception thrown of type IllegalStateException.
Please rerun your program with the --debug flag to generate a stack trace, and
file a bug report at https://github.com/kframework/k/issues
(IllegalStateException: Duplicate key UNSIGNED-INT-SYNTAX$SYNTAX (attempted
merging values module UNSIGNED-INT-SYNTAX$SYNTAX
[org.kframework.attributes.Location(Location(878,1,881,9)),
...
```
The other way I replicate it is by checking the `Definition` object directly after applying the compilation steps:
```
Kompile.java 146: org.kframework.Collections.mutable(kompiledDefinition.modules()).stream().collect(java.util.stream.Collectors.toMap(org.kframework.definition.Module::name, m -> m));
```
I tested for both java and haskell backends and I get the same problem.
The problem is pretty old. I tested the build from one year ago and it's still present.
Most likely there are multiple problematic steps since I tried removing some and I get various failure states.

**Comment by dwightguth** [comment]
I can confirm this bug affects the latest master. Also, I found a good commit in the history that predates this regression: 325fb9d75780716ede721683f2cd36e1f818d250

I am starting my bisect now and will provide the first bad commit as soon as I have it.

**Comment by radumereuta** [comment]
I found the first commit that introduces the duplicates for the java backend: 2178532dcd45dada33efcbc67e2146c6638bf91a
Leaving it here for reference.
The easiest way to track the issue is to add an assert in outer.scala in the Definition constructor:
```
case class Definition(
                       mainModule: Module,
                       entryModules: Set[Module],
                       att: Att)
  extends DefinitionToString with OuterKORE {
  assert(modules.map(x => x.name).size == modules.size) // add this line
```

**Comment by radumereuta** [comment]
I tried to enforce a more strict assert in the Module constructor:
```
assert(importedModulesIds.size == importedModuleNames.size, "Module imports duplicate modules.")
  lazy val importedModulesIds: Map[Integer, Module] = imports.map(x => (Int.box(System.identityHashCode(x)), x)).toMap ++ (imports flatMap {
    _.importedModulesIds
  })
```
but it fails really early, at the transformer from FlatModule to Module. Even after I fix that transformer, the other compilation steps fail as well. The visitor pattern implemented in scala doesn't ensure that modules don't get duplicated.

Normally it should at least ensure you won't get modules with the same name and different contents. but it can happen.
```
assert(importedModules.size == importedModuleNames.size, "Module imports duplicate modules.")
```
Since I can work around the problem and the issue is in the codebase since 2019, and it hasn't caused any other problem until now, I'm leaving it as a documented issue. We can return to it when it becomes a problem we can't work around it.

**Comment by radumereuta** [comment]
Haven't had problems with this in a while.
Probably got fixed in #1965
